reactioncommerce/reaction
# Fix 2246: Setting or changing a products perma-link causes hard refresh

Resolves #2246
Impact: minor
Type: bugfix

## Issue
Changing or creating the permalink of a product did trigger a page reload. Also this action was not reactive, meaning that other connected clients did not receive the new URL after publishing.

## Solution/Changes

- Don't trigger route change from callback value of meteor call `revisions/publish`, because
1)  it's changing server state that gets delivered to connected clients asynchronously. It's not possible to rely on the publications already being synced to the client side when the method returns. This led to race conditions when trying to change the route client side, therefore it was done the hard way by setting `window.location.href`. Tried to observe permalink changes client side, but that was unreliable as well, probably because tracker runs may re-instantiate observed cursors at any time.

2) it's not reactive (other connected clients won't get noticed about the route change), when the permalink has changed

->  INSTEAD we observe changes to handles server side and let the client know it has changed through a new property `changedHandleWas`, which will be published like any other property (gets persisted to db, but if this would be a problem we could eliminate that through usage of `Products.observe` instead of `Products.observeChanges`)

- Don't use window.location.href for route changes any more. Router.go() is used instead.

- Changes to the Product publication which seems wrong  regarding updating product.__revision    
  Don't publish the whole product object to the client. The code changed is only about the property __revisions on the product collection. Therefore it's wrong to push the whole object to the client. If done anyway, it will push product data to the client (e.g. handle, title) that is going to be updated soon after due to meteor method revision/publish, which results in **outdated** data on the client (which in turn results in template "not found" being displayed)

- Also I think it's wrong to do database queries to decide what should be pushed to the client, e.g.
```
  const revisions = Revisions.find({	
        "documentId": id,	
         "workflow.status": {	
            $nin: [	
                "revision/published"	
            ]	
          }	
        }).fetch();	
	
     fields.__revisions = revisions;	
     this.changed("Products", id, fields);
```
doesn't seem correct. To know what needs to be synced with the client, I think the source of truth needs to be the merge box (session collection view). Using db queries will not work reliably, as it seems. 

    
## Breaking changes
none expected


## Testing
1. As an admin create a new product
2. Change permalink of the product
3. Publish changes
4. See updates to content and URL without page refresh
5. URL for connected clients should update as well
